dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +122,475 @@ class FunctionVectorizedLeastGreatest : public IFunction
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        for (const auto & argument : arguments)
+        {
+            if (!argument->isString())
+            {
+                throw Exception(
+                    fmt::format("argument type not string"),
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            }
+        }
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        size_t num_arguments = arguments.size();
+        if (num_arguments < 2)
+        {
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        }
+        if (collator)
+            executeInternal<true>(block, arguments, result);
+        else
+            executeInternal<false>(block, arguments, result);
+    }
+
+    template <bool use_collator>
+    void executeInternal(Block & block, const ColumnNumbers & arguments, size_t result) const
+    {
+        size_t num_arguments = arguments.size();
+
+        using impl = StringOperationImpl<least, use_collator>;
+
+        std::vector<const ColumnConst *> const_columns;
+        std::vector<const ColumnString *> string_columns;
+        for (size_t i = 0; i < num_arguments; ++i)
+        {
+            const auto * c = block.getByPosition(arguments[i]).column.get();
+            const auto * c_string = checkAndGetColumn<ColumnString>(c);
+            const ColumnConst * c_const = checkAndGetColumnConstStringOrFixedString(c);
+            if (c_const)
+                const_columns.emplace_back(c_const);
+            if (c_string)
+                string_columns.emplace_back(c_string);
+        }
+
+        // 1. calculate result column for const columns
+        StringRef const_res;
+        if (!const_columns.empty())
+        {
+            const_res = const_columns[0]->getDataAt(0);
+            for (size_t i = 1; i < const_columns.size(); ++i)
+            {
+                StringRef b = const_columns[i]->getDataAt(0);
+                impl::constantConstant(collator, const_res, b);
+            }
+        }
+
+        if (string_columns.empty() && !const_columns.empty()) // fill result column
+        {
+            block.getByPosition(result).column
+                = block
+                      .getByPosition(result)
+                      .type->createColumnConst(const_columns[0]->size(), Field(const_res.toString()));
+            return;
+        }
+
+        // 2. calculate result column for string columns
+        auto string_columns_size = string_columns.size();
+        std::vector<StringRef> result_string_refs;
+        result_string_refs.resize(string_columns[0]->size());
```

Comment:
if `string_columns_size` is 1, this memory allocation is useless. I suggest that `result_string_refs` can be moved to `if (string_columns_size >= 2)` clause.